### PR TITLE
Created the ui for the flex deployment settings. …

### PR DESCRIPTION
**Pull request by nbashirbello**
No data persitting at this time

### CODE REVIEW

**Comment by nbashirbello** [comment]
screenshot from 2016-10-06 13 56 16

**Comment by nbashirbello** [comment]
screenshot from 2016-10-06 14 45 08

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/FlexDeployPreferencesPanel.java [hunk 1]
```diff
@@ -44,4 +58,61 @@ public boolean savePreferences() {
...
+    dockerFileText = new Text(this, SWT.LEFT | SWT.SINGLE | SWT.BORDER);
+    dockerFileBrowseButton = new Button(this, SWT.PUSH);
+    dockerFileBrowseButton.setText(Messages.getString("browse.button"));
+    dockerFileBrowseButton.addSelectionListener(new SelectionAdapter() {
+            @Override
+      public void widgetSelected(SelectionEvent event) {
+              browseForDockerFile();
+            }
+        });
+
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/FlexDeployPreferencesPanel.java]: I think you want to indent the `gaeConfigFolderLabel` and `dockerFileLabel`:
```
GridDataFactory linkedChildData = GridDataFactory.swtDefaults().indent(LINKED_CHILD_INDENT, 0);
linkedChildData.applyTo(gaeConfigFolderLabel);
linkedChildData.applyTo(dockerFileLabel);
```

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/FlexDeployPreferencesPanel.java [hunk 2]
```diff
@@ -44,4 +58,61 @@ public boolean savePreferences() {
...
+    GridLayoutFactory.fillDefaults().numColumns(3).generateLayout(this);
+  }
+
+  private void browseForConfigFolder() {
+    String last = gaeConfigFolderText.getText().trim();
+    DirectoryDialog dialog = new DirectoryDialog(getShell(), SWT.SINGLE);
+    dialog.setText(Messages.getString("config.folder.browse.button.text"));
+    dialog.setMessage(Messages.getString("config.folder.browse.button.message"));
+    dialog.setFilterPath(last);
+    String result = dialog.open();
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/FlexDeployPreferencesPanel.java]: You also need some way to convey validation status so the OK button is disabled with error text…

**Comment by briandealwis** [comment]
@nbashirbello weren't you going to Indent the text fields too?

**Comment by elharo** [comment]
Should the title be "App Engine Flex Deployment"?
What happens if the checkbox isn't checked?

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties [hunk 3]
```diff
@@ -19,6 +19,15 @@ stop.previous.version=Stop previous version
...
 use.custom.versioning=Override generated version
 
+# Flex deploy settings
+browse.button=Browse
+config.folder.browse.button.message=The App Engine configuration folder is usually called \"appengine\"
+config.folder.browse.button.text=Select the App Engine configuration folder
+config.folder.location=App Engine configuration folder:
+docker.file.browse.button.text=Select the docker file
+docker.file.location=Docker file location:
+use.config.values=Use these values during deployment
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties]: If these values have defaults, I'd prefer something similar like: "Override default values" so the wording is similar to the standard deploy UI.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesPanel.java [hunk 4]
```diff
@@ -31,4 +36,23 @@
...
 
+  @Override
+  public void dispose() {
+    if (formToolkit != null) {
+      formToolkit.dispose();
+    }
+    super.dispose();
+  }
+
+  public FormToolkit getFormToolkit() {
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesPanel.java]: protected?

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/FlexDeployPreferencesPanel.java [hunk 5]
```diff
@@ -44,4 +58,78 @@ public boolean savePreferences() {
...
+    if (result == null) {
+      return;
+    }
+    dockerFileText.setText(result);
+  }
+
+  private void updateControls() {
+    boolean enabled = useValuesButton.getSelection();
+    gaeConfigFolderLabel.setEnabled(enabled);
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/FlexDeployPreferencesPanel.java]: we could simplify this code by using:
```
for( Control child : getChildren() )
        {
            child.setEnabled( enabled );
        }
```
(from `org.eclipse.wst.common.project.facet.ui.internal.util.EnhancedComposite`)
And this was we don't need fields for the `Label` widgets.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/FlexDeployPreferencesPanel.java [hunk 6]
```diff
@@ -44,4 +58,78 @@ public boolean savePreferences() {
...
+            }
+        });
+
+    dockerFileLabel = new Label(this, SWT.LEFT);
+    dockerFileLabel.setText(Messages.getString("docker.file.location"));
+    dockerFileText = new Text(this, SWT.LEFT | SWT.SINGLE | SWT.BORDER);
+    dockerFileBrowseButton = new Button(this, SWT.PUSH);
+    dockerFileBrowseButton.setText(Messages.getString("browse.button"));
+    dockerFileBrowseButton.addSelectionListener(new SelectionAdapter() {
+            @Override
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/FlexDeployPreferencesPanel.java]: indentation is off

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/FlexDeployPreferencesPanel.java [hunk 7]
```diff
@@ -44,4 +58,78 @@ public boolean savePreferences() {
...
+            }
+        });
+
+    gaeConfigFolderLabel = new Label(this, SWT.LEFT);
+    gaeConfigFolderLabel.setText(Messages.getString("config.folder.location"));
+    gaeConfigFolderText = new Text(this, SWT.LEFT | SWT.SINGLE | SWT.BORDER);
+    gaeConfigFolderBrowseButton = new Button(this, SWT.PUSH);
+    gaeConfigFolderBrowseButton.setText(Messages.getString("browse.button"));
+    gaeConfigFolderBrowseButton.addSelectionListener(new SelectionAdapter() {
+            @Override
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/FlexDeployPreferencesPanel.java]: indentation is off

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/FlexDeployPreferencesPanel.java [hunk 8]
```diff
@@ -44,4 +58,78 @@ public boolean savePreferences() {
     return false;
   }
 
+  private void createConfigurationFilesSection() {
+    useValuesButton = new Button(this, SWT.CHECK);
+    useValuesButton.setText(Messages.getString("use.config.values"));
+    useValuesButton.setLayoutData(new GridData(SWT.LEFT, SWT.CENTER, true, false, 3, 1));
+    useValuesButton.addSelectionListener(new SelectionAdapter() {
+            @Override
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/FlexDeployPreferencesPanel.java]: indentation is off

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/FlexDeployPreferencesPanel.java [hunk 9]
```diff
@@ -44,4 +58,78 @@ public boolean savePreferences() {
...
+    dialog.setMessage(Messages.getString("config.folder.browse.button.message"));
+    dialog.setFilterPath(last);
+    String result = dialog.open();
+    if (result == null) {
+      return;
+    }
+    gaeConfigFolderText.setText(result);
+  }
+
+  private void browseForDockerFile() {
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/FlexDeployPreferencesPanel.java]: could be refactored into a single method with `browseForConfigFolder`

**Comment by nbashirbello** [comment]
I have another PR (#786) that changes the title depending on the facet

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/FlexDeployPreferencesPanel.java (hunk 9 above)
Comment (nbashirbello) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/FlexDeployPreferencesPanel.java]: The will be more differentiation code for the config folder and docker file textboxes, that's why I separated them.

**Comment by nbashirbello** [comment]
In regards to "What happens if the checkbox isn't checked?" we would not pass the extra parameters to the deploy command.

**Pull request merged**